notebooks/examples/python/BritishNationalGrid.py

```diff
@@ -0,0 +1,370 @@
+
+# COMMAND ----------
+
+# MAGIC %%mosaic_kepler
+# MAGIC to_display "geometry" "geometry(27700)" 200
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ### Performing the spatial join
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We can now do spatial join between our UPRNs and postcodes.
+
+# COMMAND ----------
+
+with_postcodes = (
+  uprns_table.join(
+    postcodes_with_index,
+    uprns_table["uprn_bng_500m"] == postcodes_with_index["mosaic_index.index_id"],
+    how = "right_outer" # to perserve even emtpy chips
+  ).where(
+    # If the borough is a core chip (the chip is fully contained within the geometry), then we do not need
+    # to perform any intersection, because any point matching the same index will certainly be contained in
+    # the borough. Otherwise we need to perform an st_contains operation on the chip geometry.
+    col("mosaic_index.is_core") | mos.st_contains(col("mosaic_index.wkb"), col("uprn_point"))
+  ).select(
+    "properties.*", "uprn_point", "UPRN", "mosaic_index.index_id", mos.st_aswkt("mosaic_index.wkb").alias("index_geometry")
+  )
+)
+
+display(with_postcodes)
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC ## Visualise the results in Kepler
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Mosaic abstracts interaction with Kepler in python through mosaic_kepler magic. <br>
+# MAGIC Mosaic_kepler magic takes care of conversion between EPSG:27700 and EPSG:4326 so that Kepler can properly render. <br>
+# MAGIC It can handle columns with bng index ids (int and str formats are both supported) and geometries that are provided in EPSG:27700. Mosaic will convert all the geometries for proper rendering. For CRSIDs that are neither EPSG:27700 nor EPSG:4326 the end user needs to convert the geometries using mos.st_transform(geom, destination_crsid).
+
+# COMMAND ----------
+
+# MAGIC %%mosaic_kepler
+# MAGIC with_postcodes "index_geometry" "geometry(27700)" 5000
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Using mosaic it takes only a few lines of code to produce BNG based heat map and visualise it in Kepler. <br>
+# MAGIC By default the colors wont be affected by the counts and you'd need to change the options in Kepler UI. <br>
+# MAGIC Navigate to the layer, expland it and for the fill color click on the 3 dots icon, then select count as the field for color scaling. 
+
+# COMMAND ----------
+
+properties_per_index = with_postcodes.groupBy("index_id").count()
+
+# COMMAND ----------
+
+# MAGIC %%mosaic_kepler
+# MAGIC properties_per_index "index_id" "bng" 6000
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC We can do the same 
+
+# COMMAND ----------
+
+properties_per_chip = with_postcodes.groupBy("index_geometry").count()
+
+# COMMAND ----------
+
+# MAGIC %md
+# MAGIC Note that if you dont use "right_outer" join some chips will be empty. <br>
+# MAGIC This is due to no UPRNs being located in that exact border part between 2 postal areas. <br>
```

Comment:
This is not clear to me. This should be valid for the full chip surface and not just the border. Why is that?

Reply:
it is actually true for any chip. I will rephrase.

Reply:
it should say if no matches for that specific cell - I just went with the fact these cases happen more often near postal area edges, but that is just the case here with london data